File: spring-boot/src/main/java/springboot/repository/FavoriteRepository.java
package springboot.repository;

import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import springboot.domain.entity.ArticleEntity;
import springboot.domain.entity.FavoriteEntity;
import springboot.domain.entity.UserEntity;

public interface FavoriteRepository extends JpaRepository<FavoriteEntity, Long> {

    Optional<FavoriteEntity> findByUserEntityAndArticleEntity(UserEntity userEntity, ArticleEntity articleEntity);

    @Query("SELECT f FROM FavoriteEntity AS f JOIN f.userEntity AS u JOIN f.articleEntity AS a WHERE u.username = :username ORDER BY a.createdAt DESC")
    Page<FavoriteEntity> findUsingUser(@Param("username") String username, Pageable pageable);
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/domain/favorite/FavoriteRepository.kt
package real.world.springbootkt.domain.favorite

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import real.world.springbootkt.domain.article.Article
import real.world.springbootkt.domain.user.User

@Repository
interface FavoriteRepository : JpaRepository<Favorite, Long> {
    fun findByUserAndArticleIn(user: User, articles: List<Article>): List<Favorite>
}